Add function apply_tuple_combinations

moxcodes opened the pull request:
## Proposed changes

Given a tuple with of length N and a function which takes m<N arguments, this
function expands to the N^m calls of the function on every combination of the
tuple elements of length m (including, for instance, the first element repeated
m times).

Two varieties provided - one which takes a std::function and infers the
argument set length, and one which requires the argument length to be specified
as a template parameter and can take any callable - so can be used with
lambdas (for which the argument length does not seem to be easily inferred.

Tests and use cases provided in `Test_Tuple.cpp`

### Types of changes:

- [ ] New feature

### Component:

- [ ] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

Similar functions can be made for other combinatoric constructs without too much effort. Happy to add them if they seem desirable.

wthrowe commented on src/Utilities/Tuple.hpp:
```diff
@@ -174,3 +174,105 @@ constexpr inline void tuple_transform(
+          Requires<sizeof...(Args) == detail::count_arg<T>::value> = nullptr>
+constexpr inline void apply_tuple_combinations(std::tuple<Elements...> /*set*/,
+                                               std::function<T> function,
+                                               Args... args) {
```
Use more references and noexcepts everywhere.

wthrowe commented on src/Utilities/Tuple.hpp:
```diff
@@ -174,3 +174,105 @@ constexpr inline void tuple_transform(
+}
+
+
+/*!
```
Use a `//@{` `//@}` doxygen grouping to avoid having two copies of this.

wthrowe commented on src/Utilities/Tuple.hpp:
```diff
@@ -174,3 +174,105 @@ constexpr inline void tuple_transform(
+ * \param function - when ArgLength is not specified, must be a
+ * std::function. If ArgLength is specified, any callable may be used.
+ *
+ * \param set - the tuple for which you'd like to execute every combination
```
Reorder to match new order of args.

wthrowe commented on src/Utilities/Tuple.hpp:
```diff
@@ -174,3 +174,105 @@ constexpr inline void tuple_transform(
+
+/*!
+ * \ingroup UtilitiesGroup
+ * \brief Perform a function over combinatoric subsets of a std::tuple
```
The documentation isn't showing up anywhere, and (possibly relatedly) there is a doxygen warning that Travis is not correctly complaining about.  (It is in the output, it just reports as passed.)

moxcodes commented:
fixed.
Cheers!

nilsdeppe commented on src/Utilities/Tuple.hpp:
```diff
@@ -174,3 +174,89 @@ constexpr inline void tuple_transform(
+/*!
+ * \brief counts the arguments of a function object.
+ */
+template <typename T>
```
We have something like this already: `tt::function_info`, you can do `tmpl::size<tt::function_info<>::argument_types>::value`

kidder replied:
ping

moxcodes replied:
ping (me) to remove

nilsdeppe commented on src/Utilities/Tuple.hpp:
```diff
@@ -174,3 +174,89 @@ constexpr inline void tuple_transform(
+template <typename T, typename... Elements, typename... Args,
+          Requires<sizeof...(Args) == detail::count_arg<T>::value> = nullptr>
+constexpr inline void apply_tuple_combinations(
+    const std::tuple<Elements...>& /*set*/, const std::function<T>& function,
```
Are you sure you need to use `std::function`? It's really bad... Has a huge overhead associated with it and it doesn't do anything useful in this PR. It's just type disambiguation if I've understood correctly.

moxcodes replied:
worse yet, just used as a convenient container to do the argument counting. Easily removed when using the function_info instead. :+1:

nilsdeppe commented on src/Utilities/Tuple.hpp:
```diff
@@ -174,3 +174,89 @@ constexpr inline void tuple_transform(
+          Requires<sizeof...(Args) == detail::count_arg<T>::value> = nullptr>
+constexpr inline void apply_tuple_combinations(
+    const std::tuple<Elements...>& /*set*/, const std::function<T>& function,
+    Args&... args) noexcept {
```
`const Args&`

nilsdeppe commented on src/Utilities/Tuple.hpp:
```diff
@@ -174,3 +174,89 @@ constexpr inline void tuple_transform(
+constexpr inline void apply_tuple_combinations(
+    const std::tuple<Elements...>& set, const std::function<T>& function,
+    Args&... args) noexcept {
+  tuple_fold(set, [&function, &set, &args...](auto x) {
```
`const auto& x` and `noexcept` the lambda

nilsdeppe commented on src/Utilities/Tuple.hpp:
```diff
@@ -174,3 +174,89 @@ constexpr inline void tuple_transform(
+};
+}  // namespace detail
+
+// compiler complains that `set' is not present due to the different ways
```
Why not just cast to void?

moxcodes replied:
sorry, I don't understand the suggestion. What I mean by this comment is that the \param set doxygen argument doesn't apply to the first definition in each pair, so generates -Wdocumentation warnings.

nilsdeppe replied:
Let me try to clarify :) I think if you uncomment `set` in those, and then do a `(void)set;` as the first line in the function the compiler will be happy without needing to use the pragmas.

moxcodes replied:
well, that's a cute trick. Thanks!

moxcodes commented:
Fixed and updated. Old commits squashed, new commit preserved.
Cheers

nilsdeppe commented on src/Utilities/Tuple.hpp:
```diff
@@ -174,3 +176,135 @@ constexpr inline void tuple_transform(
+          Requires<sizeof...(Args) == tmpl::size<typename tt::function_info<
+                                          T>::argument_types>::value> = nullptr>
+constexpr inline void apply_tuple_combinations(
+    const std::tuple<Elements...>& set, const std::function<T>& function,
```
But you're still using `std::function`. Can't you not use that?

moxcodes replied:
oops, gitfail. fixing...

kidder commented on src/Utilities/Tuple.hpp:
```diff
@@ -174,3 +174,75 @@ constexpr inline void tuple_transform(
       tuple, std::forward<N_aryOp>(op),
       std::make_index_sequence<sizeof...(Elements)>{}, args...);
 }
+
```
commit message:  "of length",  unbalanced parens in 2nd paragraph

moxcodes commented on src/Utilities/Tuple.hpp:
```diff
@@ -174,3 +176,136 @@ constexpr inline void tuple_transform(
+constexpr inline void apply_tuple_combinations(
+    const std::tuple<Elements...>& set, const T& function,
+    const Args&... args) noexcept {
+  (void)set;
```
add comment for less bewildering

moxcodes commented on src/Utilities/Tuple.hpp:
```diff
@@ -174,3 +176,136 @@ constexpr inline void tuple_transform(
+ * arguments inferrable from `tt::function_info`. If `ArgLength` is specified,
+ * any callable may be used.
+ *
+ * \param args - a set of partially assembled arguments passed recursively
```
include explanation for what happens when some args are passed in at the start

moxcodes commented on tests/Unit/Utilities/Test_Tuple.cpp:
```diff
@@ -3,13 +3,18 @@
 #include "Utilities/Gsl.hpp"
 #include "Utilities/Tuple.hpp"
 
+// IWYU pragma: no_include <bits/std_function.h>
```
probably not needed now that no std::function

moxcodes commented on tests/Unit/Utilities/Test_Tuple.cpp:
```diff
@@ -249,3 +254,81 @@ SPECTRE_TEST_CASE("Unit.Utilities.tuple_transform", "[Utilities][Unit]") {
+  // for more generic applications, with lambdas:
+  /// [tuple_combos_lambda]
+  int zero = 0;
+  double one = 1.2;
```
comment 1.2 = 1 and 2.5+ 5i = 2

moxcodes reviewed (commented):
Group code review with @markscheel @mgiesler

moxcodes commented:
Fixes applied
Cheers!

moxcodes committed:
Add apply_tuple_combinations

Add the function `apply_tuple_combinations` with recursively calls `tuple_fold`
to test several combinations of different types against the same generic
function. This is particularly useful for testing that, for instance, operators
amongst several types work as expected.

moxcodes committed:
Add addressof for tuples

Add an additional helper function for tuples, `addressof`, which populates a new
tuple with the addresses of the elements in a provided tuple.

moxcodes committed:
Add remove_nth helper function for tuples

Add the additional helper function `remove_nth` which returns a new tuple with
the Nth element omitted, where N is provided via template parameter.

moxcodes commented:
Utilities moved to narrower scope, tuple combinations simplified, and separate functionality split into separate commits.
Cheers!

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplHelper.cpp:
```diff
@@ -0,0 +1,34 @@
+  std::array<std::array<bool, 3>, 3> checklist {{{{false, false, false}},
+                                                 {{false, false, false}},
+                                                 {{false, false, false}}}};
+  auto lambda_tuple = std::make_tuple(zero, one, two);
```
I think more of these variables can be const

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplHelper.hpp:
```diff
@@ -0,0 +1,69 @@
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines functions for manipulating tuples
```
Not really so true in general and we don't require these comments, so can you just remove the `file` Doxygen comment please?

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplHelper.hpp:
```diff
@@ -0,0 +1,69 @@
+constexpr inline void apply_tuple_combinations(
+    const std::tuple<Elements...>& set, const T& function,
+    const Args&... args) noexcept {
+  (void)set; // so is used, can be documented, and no compiler warnings
```
`so it is used` (missing `it`)

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplHelper.hpp:
```diff
@@ -0,0 +1,69 @@
+ * apply the function to all ordered combinations. For instance, if the tuple
+ * contains elements (a,b,c) and the function f accepts two arguments it will be
+ * executed on all nine combinations of the three possible
+ * arguments. Importantly, the function f should then be valid for all such
```
`should` to `must` and `valid` -> `invokable`

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplHelper.hpp:
```diff
@@ -0,0 +1,69 @@
+ * arguments. Importantly, the function f should then be valid for all such
+ * argument combinations.
+ *
+ * \typaram ArgLength `apply_tuple_combinations` requires an explicit template
```
There's no second form anymore, right? Can't you use `tt::function_info` to determine `ArgLength`?

moxcodes replied:
Not if I want to use it for lambdas (which I really do). From what I understand, lambdas can't be used in any of the clever const deductions that tt:function_info uses, partially because the `auto`s in lambdas don't yield well-determined function types.

nilsdeppe replied:
Ah, I see, the reason is that in that case you don't have a function, you have a member function template. I think the documentation should read something like:
```
ArgLength the number of arguments taken by the invokable. In the case where `function` is an object or closure whose call operator is overloaded or a template it is not possible to determine the number of arguments using a type trait. One common use case would be a generic lambda. For this reason it is necessary to explicitly specify the number of arguments to the invokable.
```

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplHelper.hpp:
```diff
@@ -0,0 +1,69 @@
+ *
+ * \param set - the tuple for which you'd like to execute every combination
+ *
+ * \param function - when `ArgLength` is not specified, must be a function with
```
I don't think this is true anymore, is it?

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplHelper.hpp:
```diff
@@ -0,0 +1,69 @@
+ * arguments inferrable from `tt::function_info`. If `ArgLength` is specified,
+ * any callable may be used.
+ *
+ * \param args - a set of partially assembled arguments passed recursively. Some
```
Is this functionality still necessary? (If it is that's fine, I'm just curious)

moxcodes replied:
No, it's not strictly necessary, but it's potentially nice to have and the recursive nature of this function makes it simpler to act that way than to remove it.

nilsdeppe replied:
Okay, that's fine by me :)

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplHelper.hpp:
```diff
@@ -65,5 +65,26 @@ constexpr inline void apply_tuple_combinations(
+ * \details usage: `addr_of(make_not_null(&mytuple))`
+ */
+template <typename... ValueTypes, size_t... Is>
+decltype(auto) addr_of_impl(gsl::not_null<std::tuple<ValueTypes...>*> preset,
```
Impl should be in a detail namespace

nilsdeppe replied:
Just do `auto`

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplHelper.hpp:
```diff
@@ -65,5 +65,26 @@ constexpr inline void apply_tuple_combinations(
+}
+
+template <typename... ValueTypes>
+decltype(auto) addr_of(
```
just do `auto`

nilsdeppe replied:
Can you name this `addressof` to be consistent with the STL?

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplHelper.hpp:
```diff
@@ -65,5 +65,26 @@ constexpr inline void apply_tuple_combinations(
+template <typename... ValueTypes, size_t... Is>
+decltype(auto) addr_of_impl(gsl::not_null<std::tuple<ValueTypes...>*> preset,
+                            std::index_sequence<Is...> /*meta*/) noexcept {
+  return std::make_tuple(&(std::get<Is>(*preset))...);
```
This won't do what you think because `operator&` can be overloaded. You should use `std::addressof` instead.

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplHelper.hpp:
```diff
@@ -65,5 +65,26 @@ constexpr inline void apply_tuple_combinations(
+ * \brief given a pointer to a tuple, returns a tuple filled with pointers to
+ * the given tuples elements
+ *
+ * \details usage: `addr_of(make_not_null(&mytuple))`
```
We've been avoiding documenting functions that are trivial to call :)

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplHelper.cpp:
```diff
@@ -32,3 +32,12 @@ SPECTRE_TEST_CASE("Unit.Utilities.VectorImplHelpers_apply_tuple_combinations",
+  auto addr_test_tup = TestHelpers::VectorImpl::addr_of
+      (make_not_null(&test_tup));
+  CHECK(addr_test_tup == std::make_tuple(&std::get<0>(test_tup),
+                                       &std::get<1>(test_tup),
```
indentation here looks inconsistent

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplHelper.cpp:
```diff
@@ -32,3 +32,12 @@ SPECTRE_TEST_CASE("Unit.Utilities.VectorImplHelpers_apply_tuple_combinations",
+
+SPECTRE_TEST_CASE("Unit.DataStructures.VectorImplHelpers_TupleHelpers",
+                  "[Utilities][Unit]") {
+  auto addr_test_tup = TestHelpers::VectorImpl::addr_of
```
can this be `const`?

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplHelper.cpp:
```diff
@@ -32,3 +32,12 @@ SPECTRE_TEST_CASE("Unit.Utilities.VectorImplHelpers_apply_tuple_combinations",
   /// [tuple_combos_lambda]
 }
+
+SPECTRE_TEST_CASE("Unit.DataStructures.VectorImplHelpers_TupleHelpers",
```
Why not `.TupleHelpers` instead of `_`?

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplHelper.cpp:
```diff
@@ -0,0 +1,34 @@
+#include "Utilities/Tuple.hpp"
+#include "tests/Unit/DataStructures/Test_VectorImplHelper.hpp"
+
+SPECTRE_TEST_CASE("Unit.Utilities.VectorImplHelpers_apply_tuple_combinations",
```
Why not `.ApplyTupleCombinations` instead of `_`?

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplHelper.hpp:
```diff
@@ -86,5 +88,30 @@ decltype(auto) addr_of(
+ *
+ * \param tup the tuple from which to remove the element
+ */
+template <size_t N, typename... ValueTypes, size_t... Is1, size_t... Is2>
```
again this should be in a `detail` namespace

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplHelper.hpp:
```diff
@@ -86,5 +88,30 @@ decltype(auto) addr_of(
+}
+
+template <size_t N, typename... ValueTypes>
+decltype(auto) remove_nth(std::tuple<ValueTypes...>& tup) noexcept {
```
use a `const &` for the tuple

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplHelper.hpp:
```diff
@@ -68,50 +63,45 @@ constexpr inline void apply_tuple_combinations(
 
-// @{
+namespace detail {
+/// \cond HIDDEN_SYMBOLS
```
nothing in a detail namespace shows up in documentation, so no need for `cond`. More of a general note for you :)

nilsdeppe reviewed (commented):
I'm happy for you to squash the fixups into the appropriate commits :)

moxcodes commented:
Fixups squashed in.
Thanks for the review!

nilsdeppe reviewed (dismissed):
You're welcome. Nice job! :+1:

moxcodes commented:
Sorry, had to push one more change from doxygen parse errors on a template parameter.

nilsdeppe commented on tests/Unit/DataStructures/Test_VectorImplHelper.hpp:
```diff
@@ -60,5 +60,23 @@ constexpr inline void apply_tuple_combinations(
   });
 }
 // @}
+
```
You'll need to include `<memory>` for addressof

moxcodes replied:
I started with that, but IWYU says no. It compiles fine, so is that one you care enough about for me to pragma?

nilsdeppe replied:
Yea, according to the standard it comes from the memory header: https://en.cppreference.com/w/cpp/memory/addressof

nilsdeppe reviewed (commented):
Happy for you to squash.

@wthrowe @kidder could you please look at this when you get a chance? Thanks! :)

nilsdeppe commented:
clang-tidy is saying the includes our out of order

nilsdeppe reviewed (approved):
@wthrowe @kidder could you please look at this when you get a chance? Thanks! :)
